Fix error in flexbox alignment overflow calculations.

fantasai opened the pull request:
#### efacbbdcc42539c57f1fe9c5dcf40251ce703eef
<pre>
Fix error in flexbox alignment overflow calculations.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276382">https://bugs.webkit.org/show_bug.cgi?id=276382</a>
<a href="https://rdar.apple.com/131201271">rdar://131201271</a>

Reviewed by Alan Baradlay.

This patch fixes the logic in contentAlignmentStartOverflow correctly
clamp overflow, by handling reversed flows properly rather than returning
an overly-large value in some cases. It also fixes the error where we only
account for the last flex line.

* LayoutTests/fast/overflow/flexbox-abspos-overflow-expected.html: Added.
* LayoutTests/fast/overflow/flexbox-abspos-overflow.html: Added.
* LayoutTests/fast/overflow/flexbox-multiline-overflow-expected.html: Added.
* LayoutTests/fast/overflow/flexbox-multiline-overflow.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt: Partial revert of 273737@main
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt: Partial revert of 273737@main
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::layoutBlock): Shift reset for m_justifyContentStartOverflow out of the flex line loop.
(WebCore::contentAlignmentStartOverflow): Fix logic to correctly account for reversing.
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren): Update call to pass reversing parameter.
(WebCore::RenderFlexibleBox::alignFlexLines): Update call to pass reversing parameter, and fix multiline handling.

Canonical link: <a href="https://commits.webkit.org/280812@main">https://commits.webkit.org/280812@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/886570b3208202f3b4248fceb45337608944fb07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57700 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37028 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61322 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8145 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59828 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44664 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8333 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46750 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5772 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59730 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34726 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49858 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27575 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31498 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7149 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7149 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53449 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7421 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63003 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1614 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7496 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53970 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1620 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49869 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54083 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12768 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1365 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32857 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33943 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35027 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33688 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/fast/overflow/flexbox-abspos-overflow-expected.html:
```diff
@@ -0,0 +1,16 @@
+<html>
+<style>
+#scroller {
+	height: 200px;
```
There are tabs here.

smfr commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2202,17 +2203,19 @@ static LayoutUnit contentAlignmentStartOverflow(LayoutUnit availableFreeSpace, C
     case ContentPosition::LastBaseline:
         return 0_lu;
+    case ContentPosition::FlexStart:
+        return isReverse? -availableFreeSpace : 0_lu;
```
Space before the `?` please.

smfr commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2202,17 +2203,19 @@ static LayoutUnit contentAlignmentStartOverflow(LayoutUnit availableFreeSpace, C
-    case ContentPosition::FlexStart:
         return -availableFreeSpace;
+    case ContentPosition::FlexEnd:
+        return isReverse? 0_lu : -availableFreeSpace;
```
Ditto

smfr commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2202,17 +2203,19 @@ static LayoutUnit contentAlignmentStartOverflow(LayoutUnit availableFreeSpace, C
             || distribution == ContentDistribution::Stretch
             || distribution == ContentDistribution::SpaceBetween);
-        return -availableFreeSpace;
+        return isReverse? -availableFreeSpace : 0_lu;
```
Ditto

fantasai committed:
Fix error in flexbox alignment overflow calculations.
https://bugs.webkit.org/show_bug.cgi?id=276382
rdar://131201271

Reviewed by Alan Baradlay.

This patch fixes the logic in contentAlignmentStartOverflow correctly
clamp overflow, by handling reversed flows properly rather than returning
an overly-large value in some cases. It also fixes the error where we only
account for the last flex line.

* LayoutTests/fast/overflow/flexbox-abspos-overflow-expected.html: Added.
* LayoutTests/fast/overflow/flexbox-abspos-overflow.html: Added.
* LayoutTests/fast/overflow/flexbox-multiline-overflow-expected.html: Added.
* LayoutTests/fast/overflow/flexbox-multiline-overflow.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt: Partial revert of 273737@main
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-flexbox/negative-overflow-002-expected.txt: Partial revert of 273737@main
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::layoutBlock): Shift reset for m_justifyContentStartOverflow out of the flex line loop.
(WebCore::contentAlignmentStartOverflow): Fix logic to correctly account for reversing.
(WebCore::RenderFlexibleBox::layoutAndPlaceChildren): Update call to pass reversing parameter.
(WebCore::RenderFlexibleBox::alignFlexLines): Update call to pass reversing parameter, and fix multiline handling.

Canonical link: https://commits.webkit.org/280812@main
